File: frontend_reactnative/src/api/habits.ts
/**
 * 习惯 API
 */
import { request } from './client';
import type { Habit, HabitConfig } from '../db/types';

export async function api_getHabits(includeArchived = false): Promise<Habit[]> {
  return request<Habit[]>('/api/v1/habits/list', { include_archived: includeArchived });
}

export async function api_getHabit(id: number): Promise<Habit> {
  return request<Habit>('/api/v1/habits/get', { id });
}

export async function api_createHabit(cfg: HabitConfig): Promise<Habit> {
  return request<Habit>('/api/v1/habits/create', cfg);
}

export async function api_updateHabit(id: number, cfg: Partial<HabitConfig>): Promise<Habit> {
  return request<Habit>('/api/v1/habits/update', { id, ...cfg });
}

export async function api_deleteHabit(habit_id: number): Promise<void> {
  await request<{ ok: boolean }>('/api/v1/habits/delete', { habit_id, direction: 0 });
}

export async function api_moveHabit(habit_id: number, direction: -1 | 1): Promise<void> {
  await request<{ ok: boolean }>('/api/v1/habits/move', { habit_id, direction });
}

File: fastapi_server/app/routers/habits.py
"""习惯路由"""
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List
from app.database import get_db
from app.deps import get_current_user
from app.models.models import User
from app.schemas.habit import HabitCreate, HabitUpdate, HabitResponse, MoveRequest, HabitListRequest
from app.services import habit_service

router = APIRouter(prefix="/api/v1/habits", tags=["习惯"])


@router.post("/list", response_model=List[HabitResponse])
async def list_habits(
    body: HabitListRequest = HabitListRequest(),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await habit_service.get_habits(db, user.id, body.include_archived)


@router.post("/create", response_model=HabitResponse)
async def create_habit(
    body: HabitCreate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await habit_service.create_habit(db, user.id, body)


@router.post("/get", response_model=HabitResponse)
async def get_habit(
    body: HabitUpdate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # 用 body.id 获取
    return await habit_service.get_habit(db, body.id, user.id)


@router.post("/update", response_model=HabitResponse)
async def update_habit(
    body: HabitUpdate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await habit_service.update_habit(db, body.id, user.id, body)


@router.post("/delete")
async def delete_habit(
    body: MoveRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await habit_service.delete_habit(db, body.habit_id, user.id)
    return {"ok": True}


@router.post("/move")
async def move_habit(
    body: MoveRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await habit_service.move_habit(db, body.habit_id, user.id, body.direction)
    return {"ok": True}
